### PR TITLE
Test: added unit tests for SearchBar Component (#1886)

### LINKED ISSUE #1886: Add tests for <Search> component
Write unit tests for the `<Search>` React component to ensure expected behavior, edge case handling, and rendering logic.

## Essential Test Coverage Checklist

- [ ] **Renders successfully with minimal required props**  
- [ ] **Conditional rendering logic**  
- [ ] **Prop-based behavior** – different props affect output  
- [ ] **Event handling** – simulate user actions and verify callbacks  
- [ ] **State changes / internal logic**  
- [ ] **Default values and fallbacks**  
- [ ] **Text and content rendering**  
- [ ] **Handles edge cases and invalid inputs**  
- [ ] **Accessibility roles and labels**  
- [ ] **DOM structure / classNames / styles**

## Test Reference  
You can refer to the `AutoScrollToTop.test.tsx` file for an example of structure and best practices.  
To explore more examples, see the full component tests folder.

### PR DESCRIPTION
**Pull request by Adarsh0427**
## Proposed change

Resolves #1886 

I have added unit tests for the SearchBar component to ensure robust functionality and reliability. The test suite covers all critical aspects of the component including:

**Key Test Coverage:**
- ✅ **Component Rendering**: Tests basic rendering with minimal props and correct DOM structure
- ✅ **Conditional Logic**: Validates skeleton loading state vs input display based on `isLoaded` prop
- ✅ **User Interactions**: Comprehensive testing of input changes, focus/blur events, and clear button functionality
- ✅ **Debounced Search**: Properly tests the 750ms debounced search functionality using fake timers
- ✅ **State Management**: Verifies internal state changes and prop-driven behavior
- ✅ **Analytics Integration**: Tests GTM event tracking with correct payload structure
- ✅ **Edge Cases**: Handles empty inputs, special characters, very long strings, and rapid input changes
- ✅ **Accessibility**: Validates ARIA attributes, roles, and screen reader compatibility
- ✅ **Styling & CSS**: Ensures correct class names and DOM structure for all component states

**Changes in Search.tsx file:**
- Added `data-testid="clear-search-button"` - for identification of clear button in tests
- Added `data-testid="loading-skeleton"` - for identification of the loading skeleton component

## Checklist

- [x] I've read and followed the [contributing guidelines](https://github.com/OWASP/Nest/blob/main/CONTRIBUTING.md).
- [x] I've run `make check-test` locally; all checks and tests passed.

### CODE REVIEW

File: frontend/src/components/Search.tsx [hunk 1]
```diff
@@ -82,14 +83,15 @@ const SearchBar: React.FC<SearchProps> = ({
...
+                data-testid="clear-search-button"
                 onClick={handleClearSearch}
               >
                 <FontAwesomeIcon icon={faTimes} />
               </button>
             )}
           </>
         ) : (
-          <Skeleton className="h-12 rounded-lg" />
+          <Skeleton data-testid="loading-skeleton" className="h-12 rounded-lg" />
```
Comment (kasya) [inline frontend/src/components/Search.tsx]: @Adarsh0427 do not modify tested components. You changes should only include the test file. Check other tests to see how to get tested elements.

**Review by kasya — CHANGES_REQUESTED** [review]
These look good 👌🏼  except for one major thing ⬇️

File: frontend/src/components/Search.tsx (hunk 1 above)
Comment (Adarsh0427) [reply frontend/src/components/Search.tsx]: I will remove this attribute and fix this issue in search.test.tsx

**Comment by Adarsh0427** [comment]
I have made the changes in serach.test.tsx and search.tsx component to get tested elements without modifying the search.tsx file.
The only change I did in search.tsx is in this line 
``` 
const handleClearSearch = () => {
+  debouncedSearch.cancel()
   setSearchQuery('')
   onSearch('')
   inputRef.current?.focus()
 }
```
 as suggested by coderabbitai bot to cancel pending debounce when clearing search

File: frontend/__tests__/unit/components/Search.test.tsx [hunk 2]
```diff
@@ -287,6 +287,26 @@ describe('SearchBar Component', () => {
 
       expect(input).toHaveValue(longString)
     })
+
+    it('cancels pending debounced search when clear button is clicked', async () => {
+      const { container } = render(<SearchBar {...defaultProps} isLoaded={false} />)
+      const input = screen.getByPlaceholderText('Search projects...')
```
Comment (Adarsh0427) [inline frontend/__tests__/unit/components/Search.test.tsx:293]: added the test case to ensure that If the user types “abc” (debounced call scheduled) and immediately clicks the clear button, the scheduled callback will not fire ~750 ms later.

If the debouncedSearch.cancel() is removed from Search.tsx file
``` 
const handleClearSearch = () => {
-  debouncedSearch.cancel()
   setSearchQuery('')
   onSearch('')
   inputRef.current?.focus()
 }
```
then this test case will fail :
Screenshot 2025-08-10 at 1 54 02 PM

File: frontend/__tests__/unit/components/Search.test.tsx [hunk 3]
```diff
@@ -287,6 +287,26 @@ describe('SearchBar Component', () => {
...
+      
+      jest.advanceTimersByTime(750)
+      
+      expect(mockOnSearch).toHaveBeenCalledTimes(1)
+      expect(mockOnSearch).toHaveBeenCalledWith('') 
+      expect(mockOnSearch).not.toHaveBeenCalledWith('edge case') 
+      
+      expect(input).toHaveValue('')
+      expect(input).toHaveFocus()
+    })
```
Comment (Adarsh0427) [inline frontend/__tests__/unit/components/Search.test.tsx]: added the test case to ensure that If the user types “abc” (debounced call scheduled) and immediately clicks the clear button, the scheduled callback will not fire ~750 ms later.

If the debouncedSearch.cancel() is removed from Search.tsx file
``` 
const handleClearSearch = () => {
-  debouncedSearch.cancel()
   setSearchQuery('')
   onSearch('')
   inputRef.current?.focus()
 }
```
then this test case will fail :
Screenshot 2025-08-10 at 1 54 02 PM

**Review by kasya — APPROVED** [review]
Great updates. Thank you @Adarsh0427 !
